text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,476 @@
+because our new `VecIter` struct contains borrowed data, it needs to surface
+that fact, and the lifetime connected with it. It does so by taking an explicit
+`'vec` parameter for the relevant lifetime, and using it within.
+
+When using this struct, you can apply explicitly-named lifetimes as usual:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&'vec self) -> VecIter<'vec, T> { ... }
+}
+```
+
+However, in cases like this example, we would normally be able to leave off the
+lifetime with `&`, since there's only one source of data we could be borrowing
+from. We can do something similar with structs:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&self) -> VecIter<'_, T> { ... }
+}
+```
+
+The `'_` marker makes clear to the reader that *borrowing is happening*, which
+might not otherwise be clear.
+
+## `impl` blocks and lifetimes
+
+When writing an `impl` block for a structure that takes a lifetime parameter,
+you can give that parameter a name:
+
+```rust
+impl<T> VecIter<'vec, T> { ... }
+```
+
+This name can then be referred to in the body:
+
+```rust
+impl<T> VecIter<'vec, T> {
+    fn foo(&self) -> &'vec T { ... }
+    fn bar(&self, arg: &Bar) -> &'arg Bar { ... }
+
+    // these two are the same:
+    fn baz(&self) -> &T { ... }
+    fn baz(&self) -> &'self T { ... }
+}
+```
+
+If the type's lifetime is not relevant, you can leave it off using `'_`:
+
+```rust
+impl<T> VecIter<'_, T> { ... }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+**Note: these changes are designed to *not* require a new epoch**. They
+introduce several deprecations, which are essentially style lints. The next
+epoch should turn these deprecations into errors.
+
+## Lifetimes in `impl` headers
+
+When writing an `impl` header, it is deprecated to bind a lifetime parameter
+within the generics specification (e.g. `impl<'a>`). Instead the `impl` header
+can mention lifetimes without adding them as generics. Any lifetimes that are
+not already in scope (which, today, means any lifetime whatsoever) is treated as
+being bound as a parameter of the `impl`.
+
+Thus, where today you would write:
+
+```rust
+impl<'a> Iterator for MyIter<'a> { ... }
+impl<'a, 'b> SomeTrait<'a> for SomeType<'a, 'b> { ... }
+```
+
+tomorrow you would write:
+
+```rust
+impl Iterator for MyIter<'a> { ... }
+impl SomeTrait<'a> for SomeType<'a, 'b> { ... }
+```
+
+## Lifetimes in `fn` signatures
+
+When writing a `fn` declaration, it is deprecated to bind a lifetime parameter
+within the generics specification (e.g. `fn foo<'a>(arg: &'a str)`).
+
+Instead:
+
+- If a lifetime appears that is not already in scope, it is taken to be a new
+  binding, treated as a parameter to the function.
+- If an argument has exactly one elided lifetime, you can refer to that lifetime
+  by the argument's name preceded by `'`. Otherwise, it is not permitted to use
+  that lifetime name (unless it is already in scope, which generates a warning).
```

Comment:
Yeah, the warning is needed.
I was going to suggest a similar precaution - create lifetime names for *all* bindings in fn parameters, even if they don't have an unambiguous lifetime (or even any lifetime at all), but report an error if any of these lifetimes is actually used.